### PR TITLE
Dropout for LSTM, GRU, SimpleRNN, and Embedding layers.

### PR DESCRIPTION
**Pull request by yaringal**
Squashed commit of the following:

commit 39a59192e96fe4098f1d663384b79b10e3bcc979
Author: Yarin <yaringal@gmail.com>
Date:   Sat Feb 20 02:15:29 2016 +0000

    Squashed commit of the following:

    commit 88faa440d02df8ff356011258e3e89ce44a13e1d
    Author: Yarin <yaringal@gmail.com>
    Date:   Sat Feb 20 02:13:24 2016 +0000

        Clean up

    commit f55245199a11a202857efb1413ffa3b97c1dcfaf
    Author: Yarin <yaringal@gmail.com>
    Date:   Sat Feb 20 01:57:50 2016 +0000

        Ported dropout for LSTM, GRU, SimpleRNN, and Embedding layer to latest Keras (turned off by default).

        Squashed commit of the following:

        commit 574c4549da69f8c0831f02dce1ad05331d8b38ed
        Merge: 19ef51c bdb149d
        Author: Yarin <yaringal@gmail.com>
        Date:   Sat Feb 20 01:23:54 2016 +0000

            Merge branch 'BRNN_latest' of https://github.com/yaringal/keras into BRNN_latest

        commit 19ef51c633544f847cddebeb7a3add0936051f19
        Author: Yarin <yaringal@gmail.com>
        Date:   Sat Feb 20 01:12:23 2016 +0000

            implemented dropout in GRU and SimpleRNN

        commit bdb149d1bbff64cc6b4d694090b905153d28e33a
        Author: Yarin <yaringal@gmail.com>
        Date:   Sat Feb 20 01:12:23 2016 +0000

            implemented dropout in GRU and SimpleLSTM

        commit 72ade3f493dd725fb414cbc65a847259360be138
        Author: Yarin <yaringal@gmail.com>
        Date:   Sat Feb 20 00:52:01 2016 +0000

            clean up

        commit 9f3d213c91906b3be5c876d539819a8577bc438c
        Author: Yarin <yaringal@gmail.com>
        Date:   Sat Feb 20 00:42:58 2016 +0000

            Model test callback

        commit d4ffffc26cf24c8b7927209caad4379aac3db9c5
        Author: Yarin <yaringal@gmail.com>
        Date:   Fri Feb 19 23:47:40 2016 +0000

            removed dependence on theano

        commit 89a4e6576278564ffb882032d5a7ec5758fe00e4
        Author: Yarin <yg279@cam.ac.uk>
        Date:   Fri Feb 19 23:25:13 2016 +0000

            working BayesianLSTM and embedding dropout for theano backend

        commit 1ab4e19dfe9d49defd5575a5c2b0b880b5c46eb5
        Author: Yarin <yg279@cam.ac.uk>
        Date:   Fri Feb 19 16:41:48 2016 +0000

            working BayesianLSTM with dependence on theano

        commit 672c27401ee345a69592771cfc9ab017642b6af3
        Merge: 9360ea6 b8a9f84
        Author: Yarin <yaringal@gmail.com>
        Date:   Fri Feb 19 00:30:44 2016 +0000

            Merge https://github.com/fchollet/keras into BRNN_latest

        commit 9360ea6c25eab90e83aebb32eb187c65ed63c01d
        Author: Yarin <yaringal@gmail.com>
        Date:   Thu Feb 18 23:28:35 2016 +0000

            work in progress on BayesianLSTM

        commit b8a9f84fad1be2f27365a25b4e7f188d382d70d0
        Merge: a154495 0f3f563
        Author: François Chollet <francois.chollet@gmail.com>
        Date:   Thu Feb 18 11:24:42 2016 -0800

            Merge pull request #1756 from gw0/fix-for-refactor-callbacks

            Fix missing callback refactoring.

        commit 0f3f56327b661ade7824b18811163138755122dd
        Author: gw0 [http://gw.tnode.com/] <gw.2016@tnode.com>
        Date:   Thu Feb 18 17:01:45 2016 +0100

            Fix missing callback refactoring.

### CODE REVIEW

File: keras/backend/tensorflow_backend.py [hunk 1]
```diff
@@ -399,7 +399,7 @@ def gradients(loss, variables):
 # CONTROL FLOW
 
 def rnn(step_function, inputs, initial_states,
-        go_backwards=False, mask=None):
+        go_backwards=False, mask=None, constants=None):
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: I see no need to add a `constants` argument to the `rnn` API. I can be made part of the states (and the whole logic will be handled at the level of the `step` function of the RNN layers, as it should be...).

File: keras/layers/embeddings.py [hunk 2]
```diff
@@ -50,12 +55,13 @@ def __init__(self, input_dim, output_dim,
                  W_regularizer=None, activity_regularizer=None,
                  W_constraint=None,
                  mask_zero=False,
-                 weights=None, **kwargs):
+                 weights=None, p=0., **kwargs):
```
Comment (fchollet) [inline keras/layers/embeddings.py]: API-wise, `p` does not intuitively make sense in the context of a RNN (although it would in the context of a dropout layer). Prefer using `dropout`.

File: keras/layers/recurrent.py [hunk 3]
```diff
@@ -207,6 +233,18 @@ def build(self):
         self.W = self.init((input_dim, self.output_dim))
         self.U = self.inner_init((self.output_dim, self.output_dim))
         self.b = K.zeros((self.output_dim,))
+
+        def appendRegulariser(input_regulariser, param, regularizers_list):
```
Comment (fchollet) [inline keras/layers/recurrent.py]: Functions names should be defined in snake case.

File: keras/models.py [hunk 4]
```diff
@@ -660,6 +661,21 @@ def predict(self, X, batch_size=128, verbose=0):
         X = standardize_X(X)
         return self._predict_loop(self._predict, X, batch_size, verbose)[0]
 
+    def predict_stochastic(self, X, batch_size=128, verbose=0):
+        '''Generate stochastic output predictions for the input samples
+        batch by batch.
```
Comment (fchollet) [inline keras/models.py]: This is very descriptive of what the function does, for the layperson.

File: keras/backend/tensorflow_backend.py (hunk 1 above)
Comment (yaringal) [inline keras/backend/tensorflow_backend.py]: I attempted that at first, but it seemed to be slower and required more memory (with theano) than passing these arguments as non_sequences (constants). 
I did an experiment with a sentiment analysis dataset (10620 sequences of length 200). Time and memory per epoch were (on a Titan X GPU):
17s, 336MiB normal LSTM (original implementation)
18s, 412MiB dropout LSTM with constants (non_sequences)
21s, 638MiB dropout LSTM passing states

**Comment by fchollet** [comment]
I don't think there is sufficient justification to include the `ModelTest` callback and `predict_stochastic` method into the master codebase. I would rather have this PR focus on the addition of dropout / regularization to the RNN layers.

By the way, it seems like you have quite a few trailing spaces in your code. I suggest you install a linter such as [this one](https://pypi.python.org/pypi/pep8) to detect and remove them.

**Comment by yaringal** [comment]
Agreed

**Comment by fchollet** [comment]
Excellent. Thanks a lot, this is now merged.

I'm thinking we should plan to abstract `get_constants` in the future, since it creates a lot of code redundancy.

**Comment by yaringal** [comment]
Awesome. I was going to update the example file `imdb_lstm.py` to:
> print('Build model...')
model = Sequential()
model.add(Embedding(max_features, 128, input_length=maxlen, dropout=0.5))
model.add(LSTM(128, dropout_W=0.5, dropout_U=0.5))
model.add(Dropout(0.5))
...
model.fit(X_train, y_train, batch_size=batch_size, nb_epoch=15, validation_data=(X_test, y_test), show_accuracy=True)

This is what the results look like before dropout:
> Train on 20000 samples, validate on 5000 samples
Epoch 1/15
20000/20000 [==============================] - 42s - loss: 0.4705 - acc: 0.7728 - val_loss: 0.3841 - val_acc: 0.8310
Epoch 2/15
20000/20000 [==============================] - 42s - loss: 0.2676 - acc: 0.8935 - val_loss: 0.3634 - val_acc: 0.8368
Epoch 3/15
20000/20000 [==============================] - 42s - loss: 0.1642 - acc: 0.9410 - val_loss: 0.4106 - val_acc: 0.8336
Epoch 4/15
20000/20000 [==============================] - 42s - loss: 0.0913 - acc: 0.9700 - val_loss: 0.5181 - val_acc: 0.8234
Epoch 5/15
20000/20000 [==============================] - 42s - loss: 0.0734 - acc: 0.9762 - val_loss: 0.5570 - val_acc: 0.8300
Epoch 6/15
20000/20000 [==============================] - 42s - loss: 0.0419 - acc: 0.9865 - val_loss: 0.7032 - val_acc: 0.8236
Epoch 7/15
20000/20000 [==============================] - 42s - loss: 0.0325 - acc: 0.9891 - val_loss: 0.7419 - val_acc: 0.8260
Epoch 8/15
20000/20000 [==============================] - 42s - loss: 0.0210 - acc: 0.9935 - val_loss: 0.8792 - val_acc: 0.8264
Epoch 9/15
20000/20000 [==============================] - 42s - loss: 0.0304 - acc: 0.9897 - val_loss: 0.8041 - val_acc: 0.8174
Epoch 10/15
20000/20000 [==============================] - 42s - loss: 0.0173 - acc: 0.9953 - val_loss: 0.9233 - val_acc: 0.8240
Epoch 11/15
20000/20000 [==============================] - 42s - loss: 0.0087 - acc: 0.9974 - val_loss: 0.9507 - val_acc: 0.8318
Epoch 12/15
20000/20000 [==============================] - 42s - loss: 0.0314 - acc: 0.9906 - val_loss: 0.8402 - val_acc: 0.8174
Epoch 13/15
20000/20000 [==============================] - 42s - loss: 0.0140 - acc: 0.9953 - val_loss: 0.9607 - val_acc: 0.8238
Epoch 14/15
20000/20000 [==============================] - 42s - loss: 0.0070 - acc: 0.9980 - val_loss: 1.0835 - val_acc: 0.8260
Epoch 15/15
20000/20000 [==============================] - 42s - loss: 0.0034 - acc: 0.9988 - val_loss: 1.0937 - val_acc: 0.8174
5000/5000 [==============================] - 2s     
Test score: 1.09374942386
Test accuracy: 0.8174

And this is with dropout:
> Train on 20000 samples, validate on 5000 samples
Epoch 1/15
20000/20000 [==============================] - 45s - loss: 0.6394 - acc: 0.6253 - val_loss: 0.5157 - val_acc: 0.7760
Epoch 2/15
20000/20000 [==============================] - 45s - loss: 0.5481 - acc: 0.7310 - val_loss: 0.4186 - val_acc: 0.8206
Epoch 3/15
20000/20000 [==============================] - 45s - loss: 0.4990 - acc: 0.7649 - val_loss: 0.4269 - val_acc: 0.8170
Epoch 4/15
20000/20000 [==============================] - 45s - loss: 0.4808 - acc: 0.7808 - val_loss: 0.4056 - val_acc: 0.8182
Epoch 5/15
20000/20000 [==============================] - 45s - loss: 0.4579 - acc: 0.7937 - val_loss: 0.3991 - val_acc: 0.8176
Epoch 6/15
20000/20000 [==============================] - 45s - loss: 0.4320 - acc: 0.8108 - val_loss: 0.3816 - val_acc: 0.8308
Epoch 7/15
20000/20000 [==============================] - 45s - loss: 0.4191 - acc: 0.8171 - val_loss: 0.4286 - val_acc: 0.8154
Epoch 8/15
20000/20000 [==============================] - 45s - loss: 0.3956 - acc: 0.8334 - val_loss: 0.3768 - val_acc: 0.8316
Epoch 9/15
20000/20000 [==============================] - 45s - loss: 0.3823 - acc: 0.8355 - val_loss: 0.3767 - val_acc: 0.8336
Epoch 10/15
20000/20000 [==============================] - 45s - loss: 0.3706 - acc: 0.8433 - val_loss: 0.3679 - val_acc: 0.8348
Epoch 11/15
20000/20000 [==============================] - 45s - loss: 0.3577 - acc: 0.8482 - val_loss: 0.3702 - val_acc: 0.8328
Epoch 12/15
20000/20000 [==============================] - 45s - loss: 0.3435 - acc: 0.8543 - val_loss: 0.3741 - val_acc: 0.8338
Epoch 13/15
20000/20000 [==============================] - 45s - loss: 0.3430 - acc: 0.8552 - val_loss: 0.3705 - val_acc: 0.8302
Epoch 14/15
20000/20000 [==============================] - 45s - loss: 0.3266 - acc: 0.8627 - val_loss: 0.3655 - val_acc: 0.8422
Epoch 15/15
20000/20000 [==============================] - 45s - loss: 0.3221 - acc: 0.8648 - val_loss: 0.3704 - val_acc: 0.8402
5000/5000 [==============================] - 2s     
Test score: 0.37036307745
Test accuracy: 0.8402

How do I do this after the pull request was merged?
